Lib/test/test_shutil.py

```diff
@@ -3365,6 +3365,40 @@ def test_file2file_not_supported(self):
         finally:
             shutil._USE_CP_SENDFILE = True
 
+    def test_exception_on_enodata_call(self):
+        # Test logic when sendfile(2) call returns ENODATA error on
+        # the not-first call on the file and we need to fall back to
+        # traditional POSIX while preserving the position of where we
+        # got to in writing
+        def syscall(*args, **kwargs):
+            if not flag:
+                flag.append(None)
+                return orig_syscall(*args, **kwargs)
+            else:
+                raise OSError(errno.ENODATA, "yo")
+
+        flag = []
+        orig_syscall = eval(self.PATCHPOINT)
+        # Reduce block size so that multiple syscalls are needed
+        mock = unittest.mock.Mock()
+        mock.st_size = 65536 + 1
+        with unittest.mock.patch('os.fstat', return_value=mock) as m:
+            with unittest.mock.patch(self.PATCHPOINT, create=True,
+                                     side_effect=syscall) as m2:
+                with self.get_files() as (src, dst):
+                    with self.assertRaises(_GiveupOnFastCopy) as cm:
+                        self.zerocopy_fun(src, dst)
```

Comment:
```py
with (
    unittest.mock.patch('os.fstat', return_value=mock):
    unittest.mock.patch(self.PATCHPOINT, create=True,
                        side_effect=syscall):
    self.get_files() as (src, dst)
):
    self.assertRaises(_GiveupOnFastCopy, self.zerocopy_func, src, dst)
```

You'll need to dedent the rest of the code as well. Also, don't bind context managers if they are not used.